src/charm.py

```diff
@@ -204,50 +216,96 @@ def _on_clean_up_packages_action(self, event):
-            except Exception as e:
-                logger.error(e)
-                prefix_message = (
-                    "Clean up completed with errors, "
-                    "please refer to juju's log for details."
-                )
+        cleanup_result = clean_packages(packages_to_be_removed)
         elapsed = time.time() - start
-        logger.info("Clean up complete, took {}s".format(elapsed))
 
+        if cleanup_result is True:
+            message = "Clean up completed without errors."
+        else:
+            message = (
+                "Clean up completed with errors, please refer to juju's log for "
+                "details."
+            )
+
+        logger.info("Clean up complete, took %ds", elapsed)
         event.set_results(
             {
-                "message": "{} Freed up {}".format(prefix_message, freed_up_space),
+                "time": elapsed,
+                "message": "{} Freed up {} by cleaning {} packages".format(
+                    message, freed_up_space, len(packages_to_be_removed)
+                ),
             }
         )
 
+    def _get_mirrors(self, source: Optional[str] = None) -> List[str]:
+        """Get filtered mirrors."""
+        mirrors = list(self._stored.config.get("mirror-list", []))
+        logger.debug("found %d mirrors in charm configuration", len(mirrors))
+        if source:
+            reg_filter = re.compile(source)
+            mirrors = list(filter(reg_filter.search, mirrors))
+
+        return mirrors
+
+    def _create_tmp_apt_mirror_config(self, *mirrors) -> Path:
+        """Create tmp apt_mirror config."""
+        with NamedTemporaryFile(delete=False) as tmp_config:
+            tmp_path = Path(tmp_config.name)
+
+        config = dict(self._stored.config).copy()
+        config["mirror-list"] = mirrors
+        self._render_config(config, tmp_path)
+        return tmp_path
+
     def _on_synchronize_action(self, event):
+        """Perform synchronize action."""
         logger.info("Syncing packages")
+        start = time.time()
+        mirror_filter = event.params.get("source")
+        mirrors = self._get_mirrors(mirror_filter)
+
+        if not mirrors:
+            event.fail(
+                f"No mirror matches the filter `{mirror_filter}` or mirror-list config "
+                "options is empty. Please check mirror list with "
+                "`juju config apt-mirror mirror-list`"
+            )
+            return
+
         try:
-            start = time.time()
-            mirror_path = "{}/mirror".format(self._stored.config["base-path"])
-            for dists in Path(mirror_path).rglob("**/dists"):
-                shutil.rmtree(dists)
-            subprocess.check_output(["apt-mirror"], stderr=subprocess.STDOUT)
+            if mirror_filter is None:
+                # clean dists only if no filter was applied
+                clean_dists(Path(self._stored.config["base-path"]))
```

Comment:
I would extract this to a function

```python
def clean_dists_only_if_no_filter_was_applied(miror_filter):
    if mirror_filter is None:
        clean_dists(Path(self._stored.config["base-path"]))

```

Reply:
I do not think it's good approach to pack this to function, since there is only one function called and simple statement.

Reply:
It is not blocker but if you need to explain it, it may require function.